packages/admin-ui/src/filters/filter.state.tsx

```diff
@@ -1,59 +1,105 @@
-import { useEffect } from 'react'
-import type {
-  FilterItem,
-  Key,
-  GenericFilterStateReturn,
-} from './base-filter.state'
-import { useBaseFilterState } from './base-filter.state'
-
-export function useFilterState(props: UseFilterStateProps): UseFilterReturn {
-  const {
-    initialApplied,
-    onChange: onChangeCb = () => {},
-    ...otherProps
-  } = props
-
-  const onChange = ({ selected }: { selected: Key[] }) => {
-    onChangeCb({ selected: selected?.length ? selected[0] : null })
-  }
+import { useState, useCallback, useEffect } from 'react'
+
+import { useMenuState } from 'ariakit/menu'
+import type { ComboboxState } from '../combobox/combobox.state'
+import { useComboboxState } from '../combobox/combobox.state'
+
+export function useFilterState<T extends FilterItem>(
+  props: UseFilterStateProps<T>
+): UseFilterStateReturn<T> {
+  const { items, label, initialApplied, onChange = () => {} } = props
 
-  const filterState = useBaseFilterState({
-    ...otherProps,
-    initialApplied: initialApplied ? [initialApplied] : [],
-    onChange,
-    selectionMode: 'single',
+  const [appliedKey, setAppliedKey] = useState(initialApplied || null)
+  const [appliedItem, setAppliedItem] = useState<T>()
+
+  const combobox = useComboboxState({
+    virtualFocus: false,
+    list: items,
+    getOptionValue: (item) => item.id,
   })
 
-  const {
-    selectedKeys,
-    appliedKeys = [],
-    appliedItems = [],
-    ...singleSelectState
-  } = filterState
+  useEffect(() => {
+    const initialItem = items.find((it) => it.id === initialApplied)
+
+    combobox.setSelectedItem(initialItem)
+    setAppliedItem(initialItem)
+  }, [])
+
+  const menu = useMenuState({ ...combobox, gutter: 4 })
+
+  const apply = useCallback(() => {
+    const selected = combobox.value
+
+    setAppliedKey(selected)
+    setAppliedItem(combobox.selectedItem)
+
+    onChange({ selected })
+    menu.hide()
+  }, [onChange])
+
+  const clear = useCallback(() => {
+    setAppliedKey(null)
+    setAppliedItem(undefined)
+
+    combobox.setValue('')
+    combobox.setSelectedItem(undefined)
+
+    onChange({ selected: null })
+    menu.hide()
+  }, [onChange])
 
-  // forces apply when one item is selected
   useEffect(() => {
-    filterState.onChange()
-  }, [selectedKeys[0]])
+    // auto applies whenever a new value is selected
+    if (combobox.selectedItem) {
+      apply()
+    }
+  }, [combobox.selectedItem])
+
+  useEffect(() => {
+    // Resets combobox value when menu is closed
+    if (!menu.mounted && combobox.value) {
```

Comment:
Suggestion: similar to [this case](https://github.com/vtex/admin-ui/pull/108/files#r856450735), you could let more explicit what means this expression:
```suggestion
    if (isMenuClosed && hasValidValue) {
```